src/basic/string-util.rs

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+use std::ffi::{CStr, CString};
+use std::os::raw::c_char;
+use std::os::raw::c_int;
+use std::ptr;
+
+/// Extract the i'nth line from the specified string. Returns > 0 if there are more lines after
+/// that, and == 0 if we are looking at the last line or already beyond the last line. As special
+/// optimization, if the first line is requested and the string only consists of one line we return
+/// NULL, indicating the input string should be used as is, and avoid a memory allocation for a
+/// very common case.
+///
+/// # Safety
+///
+/// This function expects a non-NULL `ret` parameter, and will dereference it without checking.
+/// If the return value is >=0 and `*ret` is non-NULL, then it will point to a dynamically
+/// allocated buffer that the caller needs to take ownership of.
+#[no_mangle]
+pub unsafe extern "C" fn string_extract_line(
+    s: *const c_char,
+    i: usize,
+    ret: *mut *mut c_char,
+) -> c_int {
+    let s = CStr::from_ptr(s).to_string_lossy();
```

Comment:
I think, you shouldn't use `to_string_lossy` here.
2 reasons:
1. It changes the string which can lead to different meaning of command or whatever: https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.to_string_lossy
2. It can allocate which contradict to the comment `As special
 optimization, if the first line is requested and the string only consists of one line we return
 NULL, indicating the input string should be used as is, and avoid a memory allocation for a
 very common case.`

Reply:
I think `CStr::from_ptr(s).to_bytes()` should be preferred. This returns an `&[u8]` and doesn't make any assumptions about if it is valid UTF-8, which makes it more robust.